### DISCUSSION TITLE
How to trigger a delayed type: "final" state?

### DISCUSSION THREAD

**Discussion by X-Y** [statelyai/xstate#3609] [Q&A] [answered]
Newbie here. I'm making a machine that first requests a QR code, then start polling until getting a success result. It's a child machine so it would return the data after success/fail. 

My question is, is it possible to delay the final step? So I can display a success/fail message briefly before returning to the parent machine. I can probably add a intermediate state("beforeSuccess") and use "after" to achieve it, but feels like there must be a more elegant way?

      {
          id: "pollingMachine",
          initial: "showQRCode",
          context: {
            qrCode: "",
            pollingTimer: Infinity
          },
          states: {
            showQRCode: {
              invoke: {
                src: "fetchQRCode",
                onDone: {
                  target: "polling",
                  actions: assign({
                    qrCode: (context, event) => {
                      return event.data;
                    },
                    pollingTimer: (context, event) => {
                      return Date.now();
                    }
                  })
                },
                onError: {
                  target: "fail"
                }
              }
            },
            polling: {
              invoke: {
                src: "startPolling",
                onDone: {
                  target: "success",
                  cond: "hasResult"
                },
                onError: {
                  target: "fail"
                }
              },
              after: {
                1000: { target: "polling", internal: false }
              }
            },
            success: {
              type: "final"
            },
            fail: {
              type: "final",
              data: (context, event) => "failed"
            }
          }
        }

**Comment by Andarist** [comment, marked as answer]
> My question is, is it possible to delay the final step? So I can display a success/fail message briefly before returning to the parent machine. I can probably add a intermediate state("beforeSuccess") and use "after" to achieve it, but feels like there must be a more elegant way?

This is IMHO quite elegant though - it's explicit and uses the existing building blocks to achieve the result.

**Reply by X-Y** [reply]
Thanks for the reply! My concern is that this kind of delays are not really part of the business logic, wouldn't it be nicer to have it as an attribute for a state, thus de-clustering the description of a machine? 

~~I see that you can send delayed events with this function: `send({ type: 'PING' }, { to: 'pong', delay: 1000 })` but you have to label the events. Is there something similar for `onDone`?~~

Another approach I could think about is to programmatically trigger a "final" state. The docs talked about using `action.done()` but I couldn't get it to work. Do you have any examples for that(for the sake of curiosity)?

**Comment by Andarist** [comment]
> Thanks for the reply! My concern is that this kind of delays are not really part of the business logic, wouldn't it be nicer to have it as an attribute for a state, thus de-clustering the description of a machine?

It might not be part of the business process itself (it kinda depends on where you draw the line though) but it's still some behavior of your UI that you need to put **somewhere**. With a machine you at least have that documented explicitly.

Note that you could group your "almost_done" and "really_done" states in a compound state to declutter the "top-level" set of states.

> Another approach I could think about is to programmatically trigger a "final" state. The docs talked about using action.done() but I couldn't get it to work. Do you have any examples for that(for the sake of curiosity)?

This is not possible - the interpreted machine is mainly controlled by the received events. So you have to dispatch events one way or another to change the state. Note that things like `onDone`, `after` etc are still sending events internally - they are just sugar syntax for some common patterns.

**Reply by X-Y** [reply]
> Note that you could group your "almost_done" and "really_done" states in a compound state to declutter the "top-level" set of states.

I'm actually having a bit difficulty on this one. I tried the code below but it is not jumping out of the `doSuccess` state?

      success: {
              initial: "beforeSuccess",
              states: {
                beforeSuccess: {
                  after: {
                    1000: "success"
                  }
                },
                doSuccess: {
                  type: "final"
                }
              }
            },

**Reply by Andarist** [reply]
This `type: 'final'` is not the root's final state, hmm, so you would actually need smth like this:
```js
success: {
  initial: "beforeSuccess",
  states: {
    beforeSuccess: {
      after: {
        1000: "success"
      }
    },
    doSuccess: {
      type: "final"
    }
  },
  onDone: 'done'
},
done: {
  type: 'final'
}
```

**Reply by X-Y** [reply]
Looks like it's better if I use 2 top level states :) Thanks anyways!
